File: Packages/Lexicon/Sources/Lexicon/FrequencyLexicon.swift
import Foundation

public enum FrequencyLexiconError: Error, CustomStringConvertible {
    case invalidMagic(UInt32)
    case unsupportedVersion(UInt32)
    case truncated(expected: Int, actual: Int)

    public var description: String {
        switch self {
        case .invalidMagic(let m):
            return "Invalid binary format: expected magic 0x4c584331, got 0x\(String(m, radix: 16))"
        case .unsupportedVersion(let v):
            return "Unsupported version: \(v)"
        case .truncated(let expected, let actual):
            return "Truncated binary: need \(expected) bytes, file has \(actual)"
        }
    }
}

/// Binary-format unigram/bigram frequency table.
///
/// Reads the `.lex` artifact produced by `scripts/build-lexicon.py`. See
/// `Packages/Lexicon/FORMAT.md` for the full byte layout.
///
/// Memory strategy: mirrors `LemmaCore.BinaryLemmatizer` — the file is
/// memory-mapped (`Data(contentsOf:options:.alwaysMapped)`) and *never*
/// parsed into Swift collections. The initializer only reads the 32-byte
/// header and computes section offsets; every lookup does lazy, offset-based
/// reads straight out of the mapped buffer via binary search. File-backed
/// clean pages don't count against the iOS keyboard-extension jetsam limit.
///
/// String comparison is byte-exact UTF-8 (code-point order), matching the
/// Python builder's `sorted()` — never Swift `String ==`/`<`, which apply
/// Unicode canonical equivalence that the builder's raw-byte sort does not.
public final class FrequencyLexicon: Lexicon {

    private static let magic: UInt32 = 0x4C58_4331  // "LXC1" little-endian

    /// Range scans for `completions(of:)` stop after this many candidate
    /// words even if the prefix's true range is larger (only reachable with
    /// very short prefixes on a large table). See FORMAT.md.
    private static let maxCompletionScan = 20_000

    private let data: Data

    public let version: Int
    public let unigramCount: Int
    public let bigramCount: Int
    public let totalUnigramTokens: UInt64

    // Byte offsets of each section from the start of the file.
    private let stringPoolOffset: Int
    private let stringPoolSize: Int
    private let wordOffsetsOffset: Int
    private let wordLengthsOffset: Int
    private let wordFreqsOffset: Int
    private let bigramFirstIdsOffset: Int
    private let bigramSecondIdsOffset: Int
    private let bigramFreqsOffset: Int

    /// Memory-map a `.lex` file. Preferred entry point on iOS/macOS.
    public convenience init(contentsOf url: URL) throws {
        let data = try Data(contentsOf: url, options: .alwaysMapped)
        try self.init(data: data)
    }

    /// Wrap an already-loaded buffer.
    public init(data: Data) throws {
        self.data = data

        guard data.count >= 32 else {
            throw FrequencyLexiconError.truncated(expected: 32, actual: data.count)
        }

        func u32(_ byteOffset: Int) -> UInt32 {
            data.withUnsafeBytes { raw in
                raw.loadUnaligned(fromByteOffset: byteOffset, as: UInt32.self).littleEndian
            }
        }
        func u64(_ byteOffset: Int) -> UInt64 {
            data.withUnsafeBytes { raw in
                raw.loadUnaligned(fromByteOffset: byteOffset, as: UInt64.self).littleEndian
            }
        }

        let magic = u32(0)
        guard magic == Self.magic else {
            throw FrequencyLexiconError.invalidMagic(magic)
        }

        let versionRaw = u32(4)
        guard versionRaw == 1 else {
            throw FrequencyLexiconError.unsupportedVersion(versionRaw)
        }
        self.version = Int(versionRaw)

        self.unigramCount = Int(u32(8))
        self.bigramCount = Int(u32(12))
        self.stringPoolSize = Int(u32(16))
        self.totalUnigramTokens = u64(20)
        // u32 at 28 is reserved

        // Section layout — must match build-lexicon.py.
        var offset = 32

        stringPoolOffset = offset
        offset += stringPoolSize  // writer pads the pool itself to 4 bytes

        wordOffsetsOffset = offset
        offset += unigramCount * 4

        wordLengthsOffset = offset
        offset += unigramCount
        offset = (offset + 3) & ~3

        wordFreqsOffset = offset
        offset += unigramCount * 4

        bigramFirstIdsOffset = offset
        offset += bigramCount * 4

        bigramSecondIdsOffset = offset
        offset += bigramCount * 4

        bigramFreqsOffset = offset
        offset += bigramCount * 4

        guard data.count >= offset else {
            throw FrequencyLexiconError.truncated(expected: offset, actual: data.count)
        }
    }

    // MARK: - Public API

    public func frequency(of word: String) -> UInt32? {
        let key = normalizedKey(word)
        guard !key.isEmpty else { return nil }
        return withBuffer { buf in
            guard let idx = findWord(key, in: buf) else { return nil }
            return readU32(buf, at: wordFreqsOffset + idx * 4)
        }
    }

    public func bigramFrequency(_ first: String, _ second: String) -> UInt32? {
        let k1 = normalizedKey(first)
        let k2 = normalizedKey(second)
        guard !k1.isEmpty, !k2.isEmpty else { return nil }
        return withBuffer { buf in
            guard let id1 = findWord(k1, in: buf), let id2 = findWord(k2, in: buf) else { return nil }
            guard let idx = findBigram(UInt32(id1), UInt32(id2), in: buf) else { return nil }
            return readU32(buf, at: bigramFreqsOffset + idx * 4)
        }
    }

    public func completions(of prefix: String, limit: Int) -> [(word: String, frequency: UInt32)] {
        guard limit > 0 else { return [] }
        let key = normalizedKey(prefix)
        guard !key.isEmpty else { return [] }

        return withBuffer { buf in
            let lo = lowerBound(key, in: buf)
            guard lo < unigramCount else { return [] }

            let hi: Int
            if let succ = successor(key) {
                hi = lowerBound(succ, in: buf)
            } else {
                hi = unigramCount
            }
            guard hi > lo else { return [] }

            let scanEnd = min(hi, lo + Self.maxCompletionScan)
            var candidates: [(word: String, frequency: UInt32)] = []
            candidates.reserveCapacity(scanEnd - lo)
            for i in lo..<scanEnd {
                let freq = readU32(buf, at: wordFreqsOffset + i * 4)
                candidates.append((wordString(at: i, in: buf), freq))
            }
            candidates.sort {
                $0.frequency != $1.frequency ? $0.frequency > $1.frequency : $0.word < $1.word
            }
            return Array(candidates.prefix(limit))
        }
    }

    /// Raw buffer size in bytes (approximate *virtual* footprint; resident
    /// dirty memory stays near zero because the buffer is file-backed).
    public var bufferSize: Int { data.count }

    // MARK: - Internals (all operate on the mapped raw buffer)

    /// Lowercase + NFC-normalize a query word/prefix into UTF-8 bytes, the
    /// same normalization the builder applies to stored words.
    @inline(__always)
    private func normalizedKey(_ s: String) -> [UInt8] {
        Array(s.lowercased().precomposedStringWithCanonicalMapping.utf8)
    }

    @inline(__always)
    private func withBuffer<R>(_ body: (UnsafeRawBufferPointer) -> R) -> R {
        data.withUnsafeBytes(body)
    }

    @inline(__always)
    private func readU32(_ buf: UnsafeRawBufferPointer, at byteOffset: Int) -> UInt32 {
        buf.loadUnaligned(fromByteOffset: byteOffset, as: UInt32.self).littleEndian
    }

    @inline(__always)
    private func readU8(_ buf: UnsafeRawBufferPointer, at byteOffset: Int) -> UInt8 {
        buf[byteOffset]
    }

    @inline(__always)
    private func poolString(_ buf: UnsafeRawBufferPointer, offset: Int, length: Int) -> String {
        let start = stringPoolOffset + offset
        let bytes = UnsafeRawBufferPointer(rebasing: buf[start..<start + length])
        return String(decoding: bytes, as: UTF8.self)
    }

    @inline(__always)
    private func wordString(at index: Int, in buf: UnsafeRawBufferPointer) -> String {
        let offset = Int(readU32(buf, at: wordOffsetsOffset + index * 4))
        let length = Int(readU8(buf, at: wordLengthsOffset + index))
        return poolString(buf, offset: offset, length: length)
    }

    /// Lexicographic comparison of `key` against a pool string, by raw UTF-8
    /// bytes (== Unicode code-point order == the Python writer's sort
    /// order). Returns -1/0/1 like `key.compare(word)`.
    @inline(__always)
    private func compareKey(
        _ key: [UInt8], poolOffset: Int, poolLength: Int, in buf: UnsafeRawBufferPointer
    ) -> Int {
        let base = stringPoolOffset + poolOffset
        let n = min(key.count, poolLength)
        var i = 0
        while i < n {
            let a = key[i]
            let b = buf[base + i]
            if a != b { return a < b ? -1 : 1 }
            i += 1
        }
        if key.count == poolLength { return 0 }
        return key.count < poolLength ? -1 : 1
    }

    @inline(__always)
    private func compareKeyAt(_ key: [UInt8], index: Int, in buf: UnsafeRawBufferPointer) -> Int {
        let offset = Int(readU32(buf, at: wordOffsetsOffset + index * 4))
        let length = Int(readU8(buf, at: wordLengthsOffset + index))
        return compareKey(key, poolOffset: offset, poolLength: length, in: buf)
    }

    /// Exact-match binary search over the alphabetically sorted word index.
    private func findWord(_ key: [UInt8], in buf: UnsafeRawBufferPointer) -> Int? {
        var left = 0
        var right = unigramCount - 1
        while left <= right {
            let mid = (left + right) >> 1
            switch compareKeyAt(key, index: mid, in: buf) {
            case 0: return mid
            case let c where c > 0: left = mid + 1
            default: right = mid - 1
            }
        }
        return nil
    }

    /// Leftmost index `i` such that `word[i] >= key` (standard lower_bound).
    /// Used by `completions(of:)` to find the start of a prefix's range.
    private func lowerBound(_ key: [UInt8], in buf: UnsafeRawBufferPointer) -> Int {
        var lo = 0
        var hi = unigramCount
        while lo < hi {
            let mid = (lo + hi) >> 1
            // compareKey(key, word) <= 0  <=>  key <= word  <=>  word >= key
            if compareKeyAt(key, index: mid, in: buf) <= 0 {
                hi = mid
            } else {
                lo = mid + 1
            }
        }
        return lo
    }

    /// Lexicographic successor of a byte string: the smallest byte string
    /// strictly greater than every string that has `bytes` as a prefix.
    /// Returns nil if `bytes` is all-0xFF (no successor exists — treat the
    /// range as extending to the end of the table).
    private func successor(_ bytes: [UInt8]) -> [UInt8]? {
        var result = bytes
        while let last = result.last {
            if last == 0xFF {
                result.removeLast()
            } else {
                result[result.count - 1] = last + 1
                return result
            }
        }
        return nil
    }

    /// Binary search over bigrams sorted by (firstWordId, secondWordId).
    private func findBigram(_ id1: UInt32, _ id2: UInt32, in buf: UnsafeRawBufferPointer) -> Int? {
        var left = 0
        var right = bigramCount - 1
        while left <= right {
            let mid = (left + right) >> 1
            let f = readU32(buf, at: bigramFirstIdsOffset + mid * 4)
            let s = readU32(buf, at: bigramSecondIdsOffset + mid * 4)
            if f == id1 && s == id2 { return mid }
            if f < id1 || (f == id1 && s < id2) {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
        return nil
    }
}

File: AndroidClient/lib/engine/src/main/kotlin/is/solberg/lyklabord/engine/lexicon/FrequencyLexicon.kt
package `is`.solberg.lyklabord.engine.lexicon

import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.Normalizer

/** Errors thrown while opening a `.lex` file. Mirrors `FrequencyLexiconError`. */
sealed class FrequencyLexiconException(message: String) : Exception(message) {
    class InvalidMagic(magic: UInt) :
        FrequencyLexiconException("Invalid binary format: expected magic 0x4c584331, got 0x${magic.toString(16)}")

    class UnsupportedVersion(version: UInt) :
        FrequencyLexiconException("Unsupported version: $version")

    class Truncated(expected: Int, actual: Int) :
        FrequencyLexiconException("Truncated binary: need $expected bytes, file has $actual")
}

/**
 * Binary-format unigram/bigram frequency table.
 *
 * 1:1 port of `Packages/Lexicon/Sources/Lexicon/FrequencyLexicon.swift`. Reads
 * the `LXC1` v1 `.lex` artifact (see `Packages/Lexicon/FORMAT.md`). The backing
 * [ByteBuffer] is never parsed into collections; every lookup does lazy,
 * offset-based reads via binary search.
 *
 * String comparison is byte-exact UTF-8 (code-point order), matching the Python
 * builder's `sorted()` — never [String] comparison, which would apply Unicode
 * canonical equivalence the builder's raw-byte sort does not.
 *
 * The buffer must be little-endian and cover the whole file. Callers on Android
 * pass a memory-mapped `AssetManager.openFd(...)` region; tests pass a mapped or
 * heap buffer of the file bytes.
 */
class FrequencyLexicon(buffer: ByteBuffer) : Lexicon {

    private val data: ByteBuffer = buffer.duplicate().order(ByteOrder.LITTLE_ENDIAN)

    val version: Int
    val unigramCount: Int
    val bigramCount: Int
    override val totalUnigramTokens: ULong

    private val stringPoolOffset: Int
    private val stringPoolSize: Int
    private val wordOffsetsOffset: Int
    private val wordLengthsOffset: Int
    private val wordFreqsOffset: Int
    private val bigramFirstIdsOffset: Int
    private val bigramSecondIdsOffset: Int
    private val bigramFreqsOffset: Int

    init {
        val count = data.capacity()
        if (count < 32) throw FrequencyLexiconException.Truncated(32, count)

        val magicRaw = data.getInt(0).toUInt()
        if (magicRaw != MAGIC) throw FrequencyLexiconException.InvalidMagic(magicRaw)

        val versionRaw = data.getInt(4).toUInt()
        if (versionRaw != 1u) throw FrequencyLexiconException.UnsupportedVersion(versionRaw)
        version = versionRaw.toInt()

        unigramCount = data.getInt(8).toUInt().toInt()
        bigramCount = data.getInt(12).toUInt().toInt()
        stringPoolSize = data.getInt(16).toUInt().toInt()
        totalUnigramTokens = data.getLong(20).toULong()
        // u32 at 28 is reserved

        var offset = 32
        stringPoolOffset = offset
        offset += stringPoolSize // writer pads the pool itself to 4 bytes

        wordOffsetsOffset = offset
        offset += unigramCount * 4

        wordLengthsOffset = offset
        offset += unigramCount
        offset = (offset + 3) and 3.inv()

        wordFreqsOffset = offset
        offset += unigramCount * 4

        bigramFirstIdsOffset = offset
        offset += bigramCount * 4

        bigramSecondIdsOffset = offset
        offset += bigramCount * 4

        bigramFreqsOffset = offset
        offset += bigramCount * 4

        if (count < offset) throw FrequencyLexiconException.Truncated(offset, count)
    }

    // MARK: - Public API

    override fun frequency(of: String): UInt? {
        val key = normalizedKey(of)
        if (key.isEmpty()) return null
        val idx = findWord(key) ?: return null
        return readU32(wordFreqsOffset + idx * 4)
    }

    override fun bigramFrequency(first: String, second: String): UInt? {
        val k1 = normalizedKey(first)
        val k2 = normalizedKey(second)
        if (k1.isEmpty() || k2.isEmpty()) return null
        val id1 = findWord(k1) ?: return null
        val id2 = findWord(k2) ?: return null
        val idx = findBigram(id1.toUInt(), id2.toUInt()) ?: return null
        return readU32(bigramFreqsOffset + idx * 4)
    }

    override fun completions(of: String, limit: Int): List<LexEntry> {
        if (limit <= 0) return emptyList()
        val key = normalizedKey(of)
        if (key.isEmpty()) return emptyList()

        val lo = lowerBound(key)
        if (lo >= unigramCount) return emptyList()

        val succ = successor(key)
        val hi = if (succ != null) lowerBound(succ) else unigramCount
        if (hi <= lo) return emptyList()

        val scanEnd = minOf(hi, lo + MAX_COMPLETION_SCAN)
        val candidates = ArrayList<LexEntry>(scanEnd - lo)
        for (i in lo until scanEnd) {
            val freq = readU32(wordFreqsOffset + i * 4)
            candidates.add(LexEntry(wordString(i), freq))
        }
        candidates.sortWith(FREQ_DESC_THEN_WORD)
        return if (candidates.size > limit) candidates.subList(0, limit).toList() else candidates
    }

    override fun continuations(of: String, limit: Int): List<LexEntry> {
        if (limit <= 0) return emptyList()
        val key = normalizedKey(of)
        if (key.isEmpty()) return emptyList()

        val id = findWord(key) ?: return emptyList()
        val wid = id.toUInt()

        val lo = bigramLowerBound(wid)
        if (lo >= bigramCount) return emptyList()
        val hi = bigramUpperBound(wid)
        if (hi <= lo) return emptyList()

        val scanEnd = minOf(hi, lo + MAX_COMPLETION_SCAN)
        val candidates = ArrayList<LexEntry>(scanEnd - lo)
        for (i in lo until scanEnd) {
            val secondId = readU32(bigramSecondIdsOffset + i * 4).toInt()
            val freq = readU32(bigramFreqsOffset + i * 4)
            candidates.add(LexEntry(wordString(secondId), freq))
        }
        candidates.sortWith(FREQ_DESC_THEN_WORD)
        return if (candidates.size > limit) candidates.subList(0, limit).toList() else candidates
    }

    /** Raw buffer size in bytes. */
    val bufferSize: Int get() = data.capacity()

    // MARK: - Internals

    /**
     * Lowercase + NFC-normalize a query word/prefix into UTF-8 bytes, the same
     * normalization the builder applies. Folds U+2019 (curly apostrophe) to the
     * straight ASCII apostrophe, matching `build-lexicon.py`.
     */
    private fun normalizedKey(s: String): ByteArray {
        val folded = s.replace('\u2019', '\'').lowercase()
        val nfc = Normalizer.normalize(folded, Normalizer.Form.NFC)
        return nfc.toByteArray(Charsets.UTF_8)
    }

    private fun readU32(byteOffset: Int): UInt = data.getInt(byteOffset).toUInt()

    private fun readU8(byteOffset: Int): Int = data.get(byteOffset).toInt() and 0xFF

    private fun poolString(offset: Int, length: Int): String {
        val start = stringPoolOffset + offset
        val bytes = ByteArray(length)
        for (i in 0 until length) bytes[i] = data.get(start + i)
        return String(bytes, Charsets.UTF_8)
    }

    private fun wordString(index: Int): String {
        val offset = readU32(wordOffsetsOffset + index * 4).toInt()
        val length = readU8(wordLengthsOffset + index)
        return poolString(offset, length)
    }

    /**
     * Lexicographic comparison of [key] against a pool string, by raw UTF-8
     * bytes (unsigned, == Unicode code-point order). Returns -1/0/1.
     */
    private fun compareKey(key: ByteArray, poolOffset: Int, poolLength: Int): Int {
        val base = stringPoolOffset + poolOffset
        val n = minOf(key.size, poolLength)
        var i = 0
        while (i < n) {
            val a = key[i].toInt() and 0xFF
            val b = data.get(base + i).toInt() and 0xFF
            if (a != b) return if (a < b) -1 else 1
            i += 1
        }
        if (key.size == poolLength) return 0
        return if (key.size < poolLength) -1 else 1
    }

    private fun compareKeyAt(key: ByteArray, index: Int): Int {
        val offset = readU32(wordOffsetsOffset + index * 4).toInt()
        val length = readU8(wordLengthsOffset + index)
        return compareKey(key, offset, length)
    }

    /** Exact-match binary search over the alphabetically sorted word index. */
    private fun findWord(key: ByteArray): Int? {
        var left = 0
        var right = unigramCount - 1
        while (left <= right) {
            val mid = (left + right) ushr 1
            val c = compareKeyAt(key, mid)
            when {
                c == 0 -> return mid
                c > 0 -> left = mid + 1
                else -> right = mid - 1
            }
        }
        return null
    }

    /** Leftmost index i such that word[i] >= key (standard lower_bound). */
    private fun lowerBound(key: ByteArray): Int {
        var lo = 0
        var hi = unigramCount
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            if (compareKeyAt(key, mid) <= 0) hi = mid else lo = mid + 1
        }
        return lo
    }

    /**
     * Lexicographic successor of a byte string: smallest byte string strictly
     * greater than every string with [bytes] as a prefix. Null when all-0xFF.
     */
    private fun successor(bytes: ByteArray): ByteArray? {
        val result = bytes.copyOf()
        var end = result.size
        while (end > 0) {
            val last = result[end - 1].toInt() and 0xFF
            if (last == 0xFF) {
                end -= 1
            } else {
                val trimmed = result.copyOf(end)
                trimmed[end - 1] = (last + 1).toByte()
                return trimmed
            }
        }
        return null
    }

    private fun bigramLowerBound(firstId: UInt): Int {
        var lo = 0
        var hi = bigramCount
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            val f = readU32(bigramFirstIdsOffset + mid * 4)
            if (f >= firstId) hi = mid else lo = mid + 1
        }
        return lo
    }

    private fun bigramUpperBound(firstId: UInt): Int {
        var lo = 0
        var hi = bigramCount
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            val f = readU32(bigramFirstIdsOffset + mid * 4)
            if (f > firstId) hi = mid else lo = mid + 1
        }
        return lo
    }

    private fun findBigram(id1: UInt, id2: UInt): Int? {
        var left = 0
        var right = bigramCount - 1
        while (left <= right) {
            val mid = (left + right) ushr 1
            val f = readU32(bigramFirstIdsOffset + mid * 4)
            val s = readU32(bigramSecondIdsOffset + mid * 4)
            if (f == id1 && s == id2) return mid
            if (f < id1 || (f == id1 && s < id2)) left = mid + 1 else right = mid - 1
        }
        return null
    }

    // MARK: - Prefix-cursor support (see PrefixSearch.kt)

    internal fun prefixWordLength(index: Int): Int = readU8(wordLengthsOffset + index)

    internal fun prefixWordString(index: Int): String = wordString(index)

    internal fun prefixWordFrequency(index: Int): UInt = readU32(wordFreqsOffset + index * 4)

    /**
     * The UTF-8 bytes of the single Unicode scalar starting at byte [depth] of
     * word [index] (scalar length from the lead byte, clamped to the word end).
     * Callers guarantee `depth < length`.
     */
    internal fun prefixScalarBytes(index: Int, depth: Int, length: Int): ByteArray {
        val offset = readU32(wordOffsetsOffset + index * 4).toInt()
        val base = stringPoolOffset + offset
        val lead = data.get(base + depth).toInt() and 0xFF
        val scalarLength = when {
            lead < 0x80 -> 1
            lead < 0xE0 -> 2
            lead < 0xF0 -> 3
            else -> 4
        }
        val endByte = minOf(depth + scalarLength, length)
        val out = ByteArray(endByte - depth)
        for (i in depth until endByte) out[i - depth] = data.get(base + i)
        return out
    }

    /**
     * Compare word [index]'s bytes at `[depth, depth + bytes.size)` against
     * [bytes]. A word too short to cover the window sorts before it (-1).
     */
    private fun suffixCompare(index: Int, bytes: ByteArray, depth: Int): Int {
        val offset = readU32(wordOffsetsOffset + index * 4).toInt()
        val length = readU8(wordLengthsOffset + index)
        val base = stringPoolOffset + offset
        var i = 0
        while (i < bytes.size) {
            val wordIndex = depth + i
            if (wordIndex >= length) return -1
            val a = data.get(base + wordIndex).toInt() and 0xFF
            val b = bytes[i].toInt() and 0xFF
            if (a != b) return if (a < b) -1 else 1
            i += 1
        }
        return 0
    }

    /**
     * Binary search restricted to [cursor]'s range: leftmost index whose suffix
     * window compares `>= bytes` ([strict] false) or `> bytes` ([strict] true).
     */
    internal fun suffixBound(bytes: ByteArray, depth: Int, cursor: LexiconPrefixCursor, strict: Boolean): Int {
        var lo = cursor.lowerBound
        var hi = cursor.upperBound
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            val c = suffixCompare(mid, bytes, depth)
            if (if (strict) c > 0 else c >= 0) hi = mid else lo = mid + 1
        }
        return lo
    }

    companion object {
        private val MAGIC: UInt = 0x4C584331u // "LXC1" little-endian
        private const val MAX_COMPLETION_SCAN = 20_000

        private val FREQ_DESC_THEN_WORD = Comparator<LexEntry> { a, b ->
            if (a.frequency != b.frequency) {
                if (a.frequency > b.frequency) -1 else 1
            } else {
                a.word.compareTo(b.word)
            }
        }
    }
}
